text/metrics/0113-exemplars.md

```diff
@@ -0,0 +1,104 @@
+
+```
+message RawValue {
+  // Numerical value of the measurement that was recorded. Only one of these two fields is
+  // used for the data, depending on its type
+  double double_value = 0;
+  int64 int64_value = 1;
+  
+  // Exact time that the measurement was recorded
+  fixed64 time_unix_nano = 2;
+
+  // 'label:value' map of all labels that were provided by the user recording the measurement
+  repeated opentelemetry.proto.common.v1.StringKeyValue labels = 3;
+
+  // Span ID of the current trace
+  optional bytes span_id = 4;
+
+  // Trace ID of the current trace
+  optional bytes trace_id = 5;
+
+  // When sample_count is non-zero, this exemplar has been chosen in a statistically
+  // unbiased way such that the exemplar is representative of `sample_count` individual events
+  optional double sample_count = 6;
+}
+```
+
+Exemplar collection should be enabled through an optional parameter (disabled by default), and when not enabled, there should be no collection/logic performed related to exemplars. This is to ensure that when necessary, aggregators are as high performance as possible. Aggregators should also have a parameter to determine whether exemplars should only be collected if they are recorded during a sampled trace, or if tracing should have no effect on which exemplars are sampled. This allows aggregations to prioritize either the link between metrics and traces or the statistical significance of exemplars, when necessary.
+
+[#347](https://github.com/open-telemetry/opentelemetry-specification/pull/347) describes a set of standard aggregators in the metrics SDK. Here we describe how exemplars could be implemented for each aggregator.
+
+### Exemplar behaviour for standard aggregators
+
+#### HistogramAggregator
+
+The HistogramAggregator MUST (when enabled) maintain a list of exemplars whose values are distributed across all buckets of the histogram (there should be one or more exemplars in every bucket that has a population of at least one sample-able measurement). Implementations SHOULD NOT retain an unbounded number of exemplars.
+
+#### Sketch
+
+A Sketch aggregator SHOULD maintain a list of exemplars whose values are spaced out across the distribution. There is no specific number of exemplars that should be retained (although the amount SHOULD NOT be unbounded), but the implementation SHOULD pick exemplars that represent as much of the distribution as possible. (Specific details not defined, see open questions.)
+
+#### Last-Value
+
+Most (if not all) Last-Value aggregators operate asynchronously and do not ever interact with context. Since the value of a Last-Value is the last measurement (essentially the other parts of an exemplar), exemplars are not worth implementing for Last-Value.
+
+#### Exact
+
+The Exact aggregator will function by maintaining a list of `RawValue`s, which contain all of the information exemplars would carry. Therefore the Exact aggregator will not need to maintain any exemplars.
```

Comment:
Exact aggregator stores every recorded measurements, but does it also have traceid/spanid as well?

Reply:
yes since it stores them as `RawValue`s, which have a `trace_id` and `span_id` property

Reply:
Ok. I think the spec about Exact aggregator is not finalized or this is an upcoming change.

The current proto doesn't show an option to store traceid/spanid.
https://github.com/open-telemetry/opentelemetry-proto/blob/master/opentelemetry/proto/metrics/v1/metrics.proto

Reply:
My bad, I think I got your question backwards. I don't think there is any mention of Exact in the proto right now if I'm not mistaken. I realize now that it's a bit misworded, but essentially I'm trying to say in the otep that the proto would communicate Exact aggregators through a list of `RawValue`s, either through the list of exemplars on the metrics object or in some other way, but either way it would have a span + trace id.

Reply:
Continue here: https://github.com/open-telemetry/opentelemetry-proto/pull/159